cirq/sim/sampler.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Abstract base class for things sampling quantum circuits."""
+
+import abc
+from typing import (List, Union)
+
+from cirq import circuits, schedules, study
+
+
+class Sampler(metaclass=abc.ABCMeta):
+    """Something capable of sampling quantum circuits. Simulator or hardware."""
+
+    def run(
+            self,
+            program: Union[circuits.Circuit, schedules.Schedule],
```

Comment:
I'm a bit worried about this not being compatible with some important patterns for engine:
1) create program
In loop
2) create job with run context and get results

If we don't separate out 1) from 2) then people will loop on 1) and 2) and this will take a hit for latency of uploading the program.

Reply:
This interface is supposed to be identical w.r.t. the engine and the simulator, but we don't really have a concept of a saved program for the simulator. So for now I will leave it out, but we can add it later if needed.